crates/provider/src/fillers/nonce.rs

```diff
@@ -6,15 +6,89 @@ use crate::{
 use alloy_network::{Network, TransactionBuilder};
 use alloy_primitives::Address;
 use alloy_transport::{Transport, TransportResult};
+use async_trait::async_trait;
 use dashmap::DashMap;
 use futures::lock::Mutex;
 use std::sync::Arc;
 
-/// A [`TxFiller`] that fills nonces on transactions.
+/// A trait that determines the behavior of filling nonces.
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+pub trait NonceManager: Clone + Send + Sync + std::fmt::Debug + Default {
+    /// Get the next nonce for the given account.
+    async fn get_next_nonce<P, T, N>(&self, provider: &P, address: Address) -> TransportResult<u64>
+    where
+        P: Provider<T, N>,
+        N: Network,
+        T: Transport + Clone;
+}
+
+/// This [`NonceManager`] implementation will fetch the transaction count for any new account it
+/// sees.
+///
+/// Unlike [`CachedNonceManager`], this implementation does not store the transaction count locally,
+/// which results in more frequent calls to the provider, but it is more resilient to chain
+/// reorganizations.
+#[derive(Clone, Debug, Default)]
+pub struct SimpleNonceManager;
```

Comment:
```suggestion
#[non_exhaustive]
pub struct SimpleNonceManager;
```

Reply:
so you want separate methods for simple and cached managers, and the user can implement their own type and pass it to this function?
```
error[E0423]: expected value, found struct `SimpleNonceManager`
  --> crates/provider/src/fillers/nonce.rs:110:28
   |
10 |     .with_nonce_management(SimpleNonceManager)
   |                            ^^^^^^^^^^^^^^^^^^ constructor is not visible here due to private fields
```